lib/lookup/src/lookup_buf/mod.rs

```diff
@@ -0,0 +1,399 @@
+    #[instrument(level = "trace")]
+    pub fn is_valid(&self) -> Result<(), LookupError> {
+        Ok(())
+    }
+
+    #[instrument(level = "trace")]
+    pub fn clone_lookup(&self) -> Lookup {
+        Lookup::from(self)
+    }
+
+    #[instrument(level = "trace")]
+    pub fn from_str(value: &str) -> Result<LookupBuf, LookupError> {
+        Lookup::from_str(value).map(|l| l.into_buf())
+    }
+
+    pub fn from_segments(segments: Vec<SegmentBuf>) -> Self {
+        Self {
+            segments: segments.into_iter().collect(),
+        }
+    }
+
+    /// Return a borrow of the SegmentBuf set.
+    #[instrument(level = "trace")]
+    pub fn as_segments(&self) -> &VecDeque<SegmentBuf> {
+        &self.segments
+    }
+
+    /// Return the SegmentBuf set.
+    #[instrument(level = "trace")]
+    pub fn into_segments(self) -> VecDeque<SegmentBuf> {
+        self.segments
+    }
+
+    /// Merge a lookup.
+    #[instrument(level = "trace")]
+    pub fn extend(&mut self, other: Self) {
+        self.segments.extend(other.segments)
+    }
+
+    /// Returns `true` if `needle` is a prefix of the lookup.
+    #[instrument(level = "trace")]
+    pub fn starts_with(&self, needle: &LookupBuf) -> bool {
+        needle.iter().zip(&self.segments).all(|(n, s)| n == s)
+    }
+
+    /// Create the possible fields that can be followed by this lookup.
+    /// Because of coalesced paths there can be a number of different combinations.
+    /// There is the potential for this function to create a vast number of different
+    /// combinations if there are multiple coalesced segments in a path.
+    #[instrument(level = "trace")]
+    pub fn to_alternative_components(&self) -> Vec<Vec<&str>> {
+        let mut components = vec![vec![]];
+        for segment in &self.segments {
+            match segment {
+                SegmentBuf::Field(FieldBuf { name, .. }) => {
+                    for component in &mut components {
+                        component.push(name.as_str());
+                    }
+                }
+
+                SegmentBuf::Coalesce(fields) => {
+                    components = components
+                        .iter()
+                        .flat_map(|path| {
+                            fields.iter().map(move |field| {
+                                let mut path = path.clone();
+                                path.push(field.name.as_str());
+                                path
+                            })
+                        })
+                        .collect();
+                }
+
+                SegmentBuf::Index(_) => {
+                    return Vec::new();
+                }
+            }
+        }
+
+        components
+    }
+}
+
+impl FromStr for LookupBuf {
+    type Err = LookupError;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        let lookup = Lookup::from_str(input)?;
+        let lookup_buf: LookupBuf = lookup.into();
+        Ok(lookup_buf)
+    }
+}
```

Comment:
Do we want this? It seems to hide the fact that this causes an allocation to take place? If it's useful in testing, perhaps mark it with `#[cfg(test)]`?

Reply:
Although not used in Vrl code since Vrl has it's own parser for lookups, `from_str` will be used in the main Vector code, eg.

https://github.com/timberio/vector/blob/new-accessors/src/transforms/reduce/mod.rs#L85

There's some documentation about it here:

https://github.com/timberio/vector/blob/lookup-lib/lib/lookup/src/lookup_buf/mod.rs#L45-L49